src/backend/d3d12/BufferD3D12.cpp

```diff
@@ -99,16 +128,28 @@ namespace d3d12 {
         return resource->GetGPUVirtualAddress();
     }
 
+    void Buffer::OnMapReadCommandSerialFinished(uint32_t mapSerial, const void* data) {
+        CallMapReadCallback(mapSerial, NXT_BUFFER_MAP_READ_STATUS_SUCCESS, data);
+    }
+
     void Buffer::SetSubDataImpl(uint32_t start, uint32_t count, const uint32_t* data) {
-        device->GetResourceUploader()->UploadToBuffer(resource, start * sizeof(uint32_t), count * sizeof(uint32_t), reinterpret_cast<const uint8_t*>(data));
+        device->GetResourceUploader()->BufferSubData(resource, start * sizeof(uint32_t), count * sizeof(uint32_t), data);
     }
 
     void Buffer::MapReadAsyncImpl(uint32_t serial, uint32_t start, uint32_t count) {
-        // TODO(cwallez@chromium.org): Implement Map Read for the null backend
+        D3D12_RANGE readRange = { start, start + count };
+        char* data;
+        resource->Map(0, &readRange, reinterpret_cast<void**>(&data));
```

Comment:
It seems like it's possible for `resource->Map` to fail. Need to error and return appropriately if that happens. It would also be good to initialize `char* data` to `nullptr` just to avoid the possible uninitialized variable.